src/Client/ConnectionString.cpp

```diff
@@ -0,0 +1,218 @@
+#include "ConnectionString.h"
+
+#include <Common/Exception.h>
+#include <Poco/Exception.h>
+#include <Poco/URI.h>
+
+#include <array>
+#include <iostream>
+#include <string>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+}
+
+namespace
+{
+
+using namespace std::string_literals;
+using namespace std::literals::string_view_literals;
+
+constexpr auto CONNECTION_URI_SCHEME = "clickhouse:"sv;
+
+void uriDecode(std::string & uri_encoded_string, bool plus_as_space)
+{
+    std::string temp;
+    Poco::URI::decode(uri_encoded_string, temp, plus_as_space);
+    std::swap(temp, uri_encoded_string);
+}
+
+void getHostAndPort(const Poco::URI & uri, std::vector<std::vector<std::string>> & hosts_and_ports_arguments)
+{
+    auto host = uri.getHost();
+    std::vector<std::string> host_and_port;
+    if (!host.empty())
+    {
+        uriDecode(host, false);
+        host_and_port.push_back("--host="s + host);
+    }
+
+    // Port can be written without host (":9000"). Empty host name equals to default host.
+    auto port = uri.getPort();
+    if (port != 0)
+        host_and_port.push_back("--port="s + std::to_string(port));
+
+    if (!host_and_port.empty())
+        hosts_and_ports_arguments.push_back(std::move(host_and_port));
+}
+
+void getHostAndPort(
+    Poco::URI & uri,
+    std::vector<std::vector<std::string>> & hosts_and_ports_arguments,
+    std::string_view host_and_port,
+    std::string_view right_part)
+{
+    // User info does not matter in sub URI
+    auto uri_string = std::string(CONNECTION_URI_SCHEME);
+    if (!host_and_port.empty())
+    {
+        uri_string.append("//");
+        uri_string.append(host_and_port);
+    }
+
+    // Right part from string includes '/database?[params]'
+    uri_string.append(right_part);
+    try
+    {
+        uri = Poco::URI(uri_string);
+    }
+    catch (const Poco::URISyntaxException & invalid_uri_exception)
+    {
+        throw DB::Exception(DB::ErrorCodes::BAD_ARGUMENTS,
+                            "Invalid connection string syntax {}: {}", uri_string, invalid_uri_exception.what());
+    }
+
+    getHostAndPort(uri, hosts_and_ports_arguments);
+}
+
+std::string makeArgument(const std::string & connection_string_parameter_name)
+{
+    return (connection_string_parameter_name.size() == 1 ? "-"s : "--"s) + connection_string_parameter_name;
+}
+
+}
+
+namespace DB
+{
+
+bool tryParseConnectionString(
+    std::string_view connection_string,
+    std::vector<std::string> & common_arguments,
+    std::vector<std::vector<std::string>> & hosts_and_ports_arguments)
+{
+    if (!connection_string.starts_with(CONNECTION_URI_SCHEME))
+        return false;
+
+    if (connection_string.size() == CONNECTION_URI_SCHEME.size())
+        return true;
+
+    auto offset = CONNECTION_URI_SCHEME.size();
```

Comment:
(cosmetics: I am not a big fan of `auto` when it is not immediately clear from the right side what the type is. My mood would improve by 4% if I could see the concrete types.)

Reply:
Done